app/src/androidTest/java/ch/hikemate/app/ui/map/MapScreenTest.kt

```diff
@@ -166,6 +171,64 @@ class MapScreenTest : TestCase() {
     composeTestRule.onNodeWithTag(MapScreen.TEST_TAG_EMPTY_HIKES_LIST_MESSAGE).assertIsDisplayed()
   }
 
+  @Test
+  fun hikesListDisplaysHikeNameIfAvailable() {
+    // Given
+    val hikeName = "My test hike whatever"
+    val hike = HikeRoute("1", Bounds(-1.0, -1.0, 1.0, 1.0), listOf(), hikeName, null)
+    `when`(hikesRepository.getRoutes(any(), any(), any())).thenAnswer {
+      val onSuccess = it.getArgument<(List<HikeRoute>) -> Unit>(1)
+      onSuccess(listOf(hike))
+    }
+
+    // When
+    setUpMap()
+    listOfHikeRoutesViewModel.setArea(BoundingBox(0.0, 0.0, 0.0, 0.0))
+    composeTestRule.waitUntil(timeoutMillis = 5000) {
+      composeTestRule
+          .onAllNodesWithTag(MapScreen.TEST_TAG_HIKE_ITEM)
+          .fetchSemanticsNodes()
+          .isNotEmpty()
+    }
+
+    // Then
+    composeTestRule
+        .onAllNodesWithTag(TEST_TAG_HIKE_CARD_TITLE, useUnmergedTree = true)
+        .assertCountEquals(1)
+    composeTestRule
+        .onNodeWithTag(TEST_TAG_HIKE_CARD_TITLE, useUnmergedTree = true)
+        .assertTextEquals(hikeName)
+  }
+
+  @Test
+  fun hikesListDisplaysDefaultValueIfNameNotAvailable() {
+    // Given
+    val hike = HikeRoute("1", Bounds(-1.0, -1.0, 1.0, 1.0), listOf(), null, null)
+    `when`(hikesRepository.getRoutes(any(), any(), any())).thenAnswer {
+      val onSuccess = it.getArgument<(List<HikeRoute>) -> Unit>(1)
+      onSuccess(listOf(hike))
+    }
+    val context = InstrumentationRegistry.getInstrumentation().targetContext
+
+    // When
+    setUpMap()
+    listOfHikeRoutesViewModel.setArea(BoundingBox(0.0, 0.0, 0.0, 0.0))
+    composeTestRule.waitUntil(timeoutMillis = 5000) {
```

Comment:
Question: Is there another way to do that ? This would really make the tests longer in the long term

Reply:
I'm not sure what you mean with "that", are you pointing to the `waitUntil` clause? I'll look into another way to do this.

However there's probably only so much I can do, at the end of the day the test must wait until the interface is ready to be tested...

Reply:
Yes I meant the `waitUntil` (reviewing on a phone is not that good).
If you find a better way then go for it. If not, keep that

Reply:
I looked a bit into it

- `composeTestRule.awaitIdle()` and its synchronous equivalent `composeTestRule.waitForIdle()` are unreliable, tests do not pass for some reason
- `composeTestRule.waitUntil()` works but is not very clean I guess
- The solution I settled on is to use a coroutine dispatcher

I changed the tests to use an `UnconfinedTestDispatcher()`. This runs the view model's coroutines on the single main thread of the tests, meaning the test will naturally block and resume once the view model is done updating the UI (UI events update will block as well due to how `StateFlow` works).

However, this does not introduce any time gain, tests take the same time to execute (like, precisely the same time).

Reply:
So, if I understand right, the test will naturally wait for the `setArea` to finish and then do the asserts ?
If so, how is there no time gain ? In your previous version, you waited for a fixed number of seconds which would eventually be longer than your new version.

Reply:
Actually I did not wait for a fixed number of seconds, I told `composeTestRule` to `waitUntil` (condition), and that condition was that there is a hike list item in the UI. The 5000ms delay was the timeout (time after which the test should fail if the condition is still not met).

Reply:
Oh okay